File: 이것이 취업을 위한 코딩 테스트다/이진_탐색/고정점_찾기/고정점_찾기.cpp
#include <iostream>
#include <vector>

using namespace std;

int binarySearch(vector<int>& arr, int start, int end) {
  if (start > end) return -1;
  int mid = (start + end) / 2;
  if (arr[mid] == mid) return mid;
  else if (arr[mid] > mid) return binarySearch(arr, start, mid - 1);
  else return binarySearch(arr, mid + 1, end);
}

int n;
vector<int> arr;

int main(void) {
  cin >> n;
  for (int i = 0; i < n; i += 1) {
    int x;
    cin >> x;
    arr.push_back(x);
  }

  int index = binarySearch(arr, 0, n - 1);
  cout << index << endl;
}

File: 이것이 취업을 위한 코딩 테스트다/이진_탐색/고정점_찾기/고정점_찾기.py
# n = int(input())
# arr = list(map(int, input().split()))

# start = 0
# end = len(arr) - 1
# answer = -1

# while start <= end:
#   mid = (start + end) // 2

#   if mid == arr[mid]:
#     answer = mid
#     break;
  
#   if arr[mid] < mid:
#     start = mid + 1
  
#   if arr[mid] > mid:
#     end = mid - 1
    
# print(answer)

def binary_search(array, start, end):
  if start> end:
    return None
  mid = (start + end) // 2
  if array[mid] == mid:
    return mid
  elif array[mid] > mid:
    return binary_search(array, start, mid - 1)
  else:
    return binary_search(array, mid + 1, end)

n = int(input())
array = list(map(int, input().split()))

index = binary_search(array, 0, n - 1)

if index == None:
  print(-1)
else:
  print(index)